src/dmd/backend/go.d

```diff
@@ -0,0 +1,500 @@
+    MFcse   = 0x10,            // global common subexpressions
+    MFvbe   = 0x20,            // very busy expressions
+    MFtime  = 0x40,            // favor time (speed) over space
+    MFli    = 0x80,            // loop invariants
+    MFliv   = 0x100,           // loop induction variables
+    MFcp    = 0x200,           // copy propagation
+    MFcnp   = 0x400,           // constant propagation
+    MFloop  = 0x800,           // loop till no more changes
+    MFtree  = 0x1000,          // optelem (tree optimization)
+    MFlocal = 0x2000,          // localize expressions
+    MFall   = 0xFFFF,          // do everything
+}
+
+/**********************************
+ * Definition elem vector, used for reaching definitions.
+ */
+
+struct DefNode
+{
+    elem    *DNelem;        // pointer to definition elem
+    block   *DNblock;       // pointer to block that the elem is in
+    vec_t    DNunambig;     // vector of unambiguous definitions
+}
+
+/* Global Variables */
+//extern __gshared uint[] optab;
+
+/* Global Optimizer variables
+ */
+struct GlobalOptimizer
+{
+    mftype mfoptim;
+    uint changes;       // # of optimizations performed
+
+    DefNode *defnod;    // array of definition elems
+    uint deftop;        // # of entries in defnod[]
+    uint defmax;        // capacity of defnod[]
+    uint unambigtop;    // number of unambiguous defininitions ( <= deftop )
+
+    vec_base_t *dnunambig;  // pool to allocate DNunambig vectors from
+    uint    dnunambigmax;   // capacity of dnunambig[]
+
+    elem **expnod;      // array of expression elems
+    uint exptop;        // top of expnod[]
+    block **expblk;     // parallel array of block pointers
+
+    vec_t defkill;      // vector of AEs killed by an ambiguous definition
+    vec_t starkill;     // vector of AEs killed by a definition of something that somebody could be
+                        // pointing to
+    vec_t vptrkill;     // vector of AEs killed by an access
+}
+
+extern __gshared GlobalOptimizer go;
+
+/* gdag.c */
+void builddags();
+void boolopt();
+void opt_arraybounds();
+
+/* gflow.c */
+void flowrd();
+void flowlv();
+void flowae();
+void flowvbe();
+void flowcp();
+void flowae();
+void genkillae();
+void flowarraybounds();
+int ae_field_affect(elem *lvalue,elem *e);
+
+/* glocal.c */
+void localize();
+
+/* gloop.c */
+int blockinit();
+void compdom();
+void loopopt();
+void fillInDNunambig(vec_t v, elem *e);
+void updaterd(elem *n,vec_t GEN,vec_t KILL);
+
+/* gother.c */
+void rd_arraybounds();
+void rd_free();
+void constprop();
+void copyprop();
+void rmdeadass();
+void elimass(elem *);
+void deadvar();
+void verybusyexp();
+list_t listrds(vec_t, elem *, vec_t);
+
+/* gslice.c */
+void sliceStructs();
+
+/***************************************************************************/
+
+extern (C++) void mem_free(void* p);
+
+void go_term()
+{
+    vec_free(go.defkill);
+    vec_free(go.starkill);
+    vec_free(go.vptrkill);
+version (Posix)
+{
```

Comment:
Seems to be some unusual indentations in some places, like this one.

Reply:
I was trying to minimize diffs with the C version.

Reply:
I suspected that.

Reply:
Ah, looking at the individual commits worked a lot better, now it makes more sense. 👍